db-copy.cpp

```diff
@@ -28,9 +28,18 @@ db_copy_thread_t::~db_copy_thread_t() { finish(); }
 void db_copy_thread_t::add_buffer(std::unique_ptr<db_cmd_t> &&buffer)
 {
     assert(m_worker.joinable()); // thread must not have been finished
-    std::unique_lock<std::mutex> lock(m_queue_mutex);
-    m_worker_queue.push_back(std::move(buffer));
-    m_queue_cond.notify_one();
+
+    for (;;) {
+        std::unique_lock<std::mutex> lock(m_queue_mutex);
+        if (m_worker_queue.size() >= db_cmd_copy_t::Max_buffers) {
+            m_queue_full_cond.wait(lock);
```

Comment:
just checked how this is done in _c++ concurrency in action_. They use a lambda function for the condition variable wait method: https://github.com/anthonywilliams/ccia_code_samples/blob/master/listings/listing_4.1.cpp#L47 (code example is not about a bounded queue, it's just to show how they use condition variables throughout chapter 4).

Reply:
Changed that for both condition variables.